BUILDING.md

```diff
@@ -0,0 +1,128 @@
+
+### `build-tools`
+
+This project contains all build logic that we publish for third party elasticsearch plugin authors.
+We provide the following plugins:
+
+- `elasticsearch.esplugin` - A gradle plugin for building an elasticsearch plugin.
+- `elasticsearch.testclusters` - A gradle plugin for setting up es clusters for testing within a build.
+
+This project is published as part of the elasticsearch release and accessible by
+`org.elasticsearch.gradle:build-tools:<versionNumber>`.
+These build tools are also used by the `elasticsearch-hadoop` project maintained by elastic.
+
+### `build-tools-internal`
+
+This project contains all elasticsearch project specific build logic that is not meant to be shared
+with other internal or external projects.
+
+## Build guidelines
+
+This is an intentionally small set of guidelines to build users and authors
+to ensure we keep the build consistent. We also publish elasticsearch build logic
+as `build-tools` to be usuable by thirdparty elasticsearch plugin authors. This is
+also used by other elastic teams like `elasticsearch-hadoop`.
+Breaking changes should therefore be avoided and an appropriate deprecation cycle
+should be followed.
+
+### Stay up to date
+
+The elasticsearch build usually uses the latest Gradle GA release. We stay as close to the
+latest Gradle releases as possible. In certain cases an update is blocked by a breaking behaviour
+in Gradle. We're usually in contact with the gradle team here or working on a fix
+in our build logic to resolve this.
+
+<b>The elasticsearch build fails the build if there's any deprecated gradle api used.</b>
+
+### Make a change in the build
+
+There are a few guidelines to follow that should make your life easier to make changes to the elasticsearch build.
+Please add a member of the `es-delivery` team as a reviewer if you're making non trivial changes build.
+
+#### Custom Plugin and Task implementations
+
+Build logic that is used across multiple subprojects should considered to be moved into a Gradle plugin with according Gradle task implmentation.
+Elasticsearch specific build logic is located in the `build-tools-internal` subproject including integration tests.
+
+- Gradle plugins and Tasks should be written in Java
+- We use a groovy and spock for setting up Gradle integration tests.
+  (see https://github.com/elastic/elasticsearch/blob/master/build-tools/src/testFixtures/groovy/org/elasticsearch/gradle/fixtures/AbstractGradleFuncTest.groovy)
+
+#### Declaring tasks
+
+The elasticsearch build makes use of the [task avoidance API](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html) to keep the configuration time of the build low.
+
+- When declaring tasks (in build scripts or custom plugins) this means that we want to _register_ a task like:
+
+
+  a task should be declared via
+
+  ```
+  tasks.register('someTask') {
+  }
+  ```
+
+  instead of eagerly _creating_ the task
+  ```
+  task someTask {
+  }
+  ```
```

Comment:
```suggestion
When declaring tasks (in build scripts or custom plugins) this means that we want to _register_ a task like:

    tasks.register('someTask') { ... }

instead of eagerly _creating_ the task:

    task someTask { ... }
```